src/geckomat/change_model/findMetSmiles.m

```diff
@@ -66,32 +66,35 @@
                 %keep the first suggestion.
                 smileResult       = regexp(smileResult,'(^\S*)\n','once','tokens');
                 uniqueSmiles{i,1} = smileResult{1,1};
-                % Append one line each time, in case internet connection is lost
-                % halfway. Open & close file each time to avoid leaving the file
-                % open when breaking the function.
-                out = [uniqueNames(i), uniqueSmiles(i)];
-                fID = fopen(smilesDBfile,'a');
-                fprintf(fID,'%s\t%s\n',out{:});
-                fclose(fID);
-                break % success? look at next metabolite
+                retry = 15; % success: no retry
             catch exception
                 %Sometimes the call fails, for example since the server is busy. In those cases
                 %we will try 10 times. Some errors however are because the metabolite
                 %name does no exist in the database (404) or some other error (the metabolite contains
                 %a slash or similar, 400 or 500). In those cases we can
                 %immediately give up.
                 if (strcmp(exception.identifier, 'MATLAB:webservices:HTTP404StatusCodeError') || ...
-                    strcmp(exception.identifier, 'MATLAB:webservices:HTTP400StatusCodeError') || ...
-                    strcmp(exception.identifier, 'MATLAB:webservices:HTTP500StatusCodeError'))  
-                    break
+                        strcmp(exception.identifier, 'MATLAB:webservices:HTTP400StatusCodeError') || ...
+                        strcmp(exception.identifier, 'MATLAB:webservices:HTTP500StatusCodeError'))
+                    uniqueSmiles(i) = {''};
+                    retry = 15;
                 else
                     retry = retry + 1;
                 end
             end
+        if retry == 10
             error('Cannot reach PubChem. Check your internet connection and try again.')
         end
+        end
+        % Append one line each time, in case internet connection is lost
+        % halfway. Open & close file each time to avoid leaving the file
+        % open when breaking the function.
+        out = [uniqueNames(i), uniqueSmiles(i)];
+        fID = fopen(smilesDBfile,'a');
+        fprintf(fID,'%s\t%s\n',out{:});
+        fclose(fID);
```

Comment:
I remember this code struck me the first time I saw it. To my understanding, the aim here is to handle situations when the connection drops while this function is running. Why does the input need to be written to file every line? how about using a temporary data structure to hold all the entries, and do the writing in one go? It would save some time having to open and close a file several thousand times.

Also, while it looks like the core was built to handle mid-way interruptions, neither the documentation nor the output make this clear. It would be beneficial for the users to know that X entries were retrieved from the file before the querying starts.

Reply:
Good suggestions, but I don't have the resource to refactor this at the moment.